### PR TITLE
Refactor Process teardown

### PR DESCRIPTION
**Pull request by sporksmith**
Refactors Process to remove more state variables and make process teardown easier to follow.

### CODE REVIEW

File: src/main/host/thread.rs [hunk 1]
```diff
@@ -13,12 +13,19 @@ use shadow_shmem::allocator::{Allocator, ShMemBlock};
...
 use crate::cshadow as c;
 use crate::host::syscall_condition::{SysCallConditionRef, SysCallConditionRefMut};
 use crate::utility::{syscall, IsSend, SendPointer};
 
+#[derive(Debug)]
+pub enum ResumeResult {
+    Blocked,
+    ExitedThread,
+    ExitedProcess,
+}
```
Comment (stevenengler) [inline src/main/host/thread.rs]: This could use some comments.

File: src/main/host/managed_thread.rs [hunk 2]
```diff
@@ -25,6 +25,13 @@ use crate::cshadow;
 use crate::host::syscall_types::SyscallReturn;
 use crate::utility::{childpid_watcher, pod, syscall};
 
+#[derive(Debug)]
+pub enum ResumeResult {
+    Blocked(SysCallCondition),
+    ExitedThread,
+    ExitedProcess,
+}
```
Comment (stevenengler) [inline src/main/host/managed_thread.rs]: This could use some comments.

File: src/main/host/managed_thread.rs [hunk 3]
```diff
@@ -96,10 +100,11 @@ impl ManagedThread {
...
             Err(SelfContainedChannelError::WriterIsClosed) => {
-                trace!("Plugin exited while executing native syscall {n}");
-                ctx.process.mark_as_exiting();
-                self.cleanup_after_exit_initiated();
-                SysCallReg::from(-(Errno::ESRCH as i64))
+                // This should only happen if the syscall was SYS_exit or
+                // SYS_exit_group, or if the process was killed externally, e.g.
+                // by sending a signal from outside of shadow. These cases would be
+                // tricky to support, and we shouldn't need to.
+                panic!("Plugin exited while executing native syscall {n}");
```
Comment (stevenengler) [inline src/main/host/managed_thread.rs:107]: Is it easy to get the exit status here and include it in the panic message? This might happen in practice if the oom manager kills some plugin process, which might be hard for the user to debug.

File: src/main/host/process.rs [hunk 4]
```diff
@@ -545,6 +544,11 @@ impl Process {
     }
 
     pub fn stop(&self, host: &Host) {
+        if !self.is_running() {
+            info!("process {} has already stopped", self.name());
+            return;
+        }
```
Comment (stevenengler) [inline src/main/host/process.rs:555]: nit: make this debug instead since it doesn't seem relevant to the user?

**Review by stevenengler — APPROVED** [review]

File: src/main/host/process.rs (hunk 4 above)
Comment (sporksmith) [reply src/main/host/process.rs]: done
